### ISSUE TITLE
Problem with reliably pulling docker containers after setting up docker

### ISSUE THREAD

**Issue by LarsFronius** [saltstack-formulas/docker-formula#71]
pillar data

```
{%- from "foobar/shared_vars.jinja" import shared_vars with context %}
docker-containers:
  lookup:
    ecs_agent:
      image: "amazon/amazon-ecs-agent:v1.7.1"
      runoptions:
        - "--detach=true"
        - "--restart=on-failure:10"
        - "--volume=/var/run/docker.sock:/var/run/docker.sock"
        - "--volume=/var/log/ecs/:/log"
        - "--volume=/var/lib/ecs/data:/data"
        - "--volume=/sys/fs/cgroup:/sys/fs/cgroup:ro"
        - "--volume=/var/run/docker/execdriver/native:/var/lib/docker/execdriver/native:ro"
        - "--publish=127.0.0.1:51678:51678"
        - "--env=ECS_LOGFILE=/log/ecs-agent.log"
        - "--env=ECS_LOGLEVEL=info"
        - "--env=ECS_DATADIR=/data"
        - "--env=ECS_CLUSTER={{ grains.tags.cluster }}"
{% endif %}

docker-pkg:
  lookup:
    config:
      - DOCKER_OPTS="{{ shared_vars.docker_opts }}"
    version: {{ shared_vars.docker_engine_version }}
    install_docker_py: False
```

state data

```
include:
  - docker
  - docker.containers
```

what we sometimes see in the logs:

```
[INFO    ] Completed state [docker] at time 10:17:12.994250
[INFO    ] Running state [docker pull amazon/amazon-ecs-agent:v1.7.1] at time 10:17:12.995713
[INFO    ] Executing state cmd.run for docker pull amazon/amazon-ecs-agent:v1.7.1
[INFO    ] Executing command 'docker pull amazon/amazon-ecs-agent:v1.7.1' in directory '/root'
[ERROR   ] Command 'docker pull amazon/amazon-ecs-agent:v1.7.1' failed with return code: 1
[ERROR   ] stdout: Warning: failed to get default registry endpoint from daemon (Cannot connect to the Docker daemon. Is the docker daemon running on this host?). Using system default: https://index.docker.io/v1/
[ERROR   ] stderr: Cannot connect to the Docker daemon. Is the docker daemon running on this host?
[ERROR   ] retcode: 1
[ERROR   ] {'pid': 5775, 'retcode': 1, 'stderr': 'Cannot connect to the Docker daemon. Is the docker daemon running on this host?', 'stdout': 'Warning: failed to get default registry endpoint from daemon (Cannot connect to the Docker daemon. Is the docker daemon running on this host?). Using system default: https://index.docker.io/v1/'}
[INFO    ] Completed state [docker pull amazon/amazon-ecs-agent:v1.7.1] at time 10:17:13.046047
```

It works mostly when I run salt again. so there seems to be a bit of time after docker is set up where the daemon is not reachable which imho the salt module should wait for.

**Comment by gravyboat** [comment]
Did you create an issue over in the main line salt repo? Not sure there's anything we can do here other than to put in a wait.

**Comment by LarsFronius** [comment]
I wonder if there is a way to reliably know when the docker daemon is able to pull images and thus wait for that. Do you know of one or how this is approached in other cfg mgmts?

what do you mean by the main line salt repo? the one where actually salt is being developed and what does that have to do with the docker-formula over here? as far as I can see it fails on a plain "docker pull" cmd.

**Comment by gravyboat** [comment]
Oh jeez, I just reviewed the formula and realized it isn't using the new dockerng stuff, this is pretty problematic as I THINK the new docker stuff avoids some of these issues, I think this is going to be EoL'd at some point as well. @LarsFronius can you take a look at https://github.com/gravyboat/docka-docka-docka/tree/master/docka-salt/srv/salt/docker and see if it does what you need in a more sensible way?

**Comment by LarsFronius** [comment]
Yeah exactly :)
I'll have a look tomorrow (CET) to maybe just rewrite the container part here towards the dockerng stuff provided by salt, alright? If that gets the job done, we killed two "issues" in one go.

What's the status actually with testing saltstack formulas? Could we come up with an easy way to make sure this formula works on redhat/debian "always" the way we expect it to be? Since docker is moving pretty fast I think its even more important to try to keep this formula at a stable level.

**Comment by gravyboat** [comment]
That would be awesome! Right now there isn't any testing past what individuals do, we've discussed some options in the past regarding using travis to test all the formulas which I would love to see, but have been lazy about implementing.

**Comment by noelmcloughlin** [comment]
Resolved by #137
